### ISSUE TITLE
Audio capture: the state or buffer may be wrong if there is a pause or other trigger

### ISSUE THREAD

**Issue by joey-cyt** [libsdl-org/SDL#9700]
When performing audio capture/playback, sometimes the playbacked audio adds noise or distortion. It's difficult to determine when and how the issue is triggered; the program may run for a whole day without encountering this problem, while at other times, the issue may start to appear after a couple of hours. Once the noise/distortion begins, the way to return to normal is to close the input and then reopen it (please refer to the following test code), or to restart the program.

I have written a simple test program as follows. When the program starts, it plays sound from the default input nicely. If debug break the while loop, for example, at the line "if (false)," and then continues, the sound will have noise or distortion. I'm not sure if debug break trigger the same issue as our application does after a long run, but the symptoms are the same. If, after debug break,  move the running line to inside "if (false){}" to close/open the input, the issue will go away.

The test is on Windows 11 x64, with main branch code at 4/30.
Attached is a recorded example of the problematic audio.
[SDL-sound-noise.zip](https://github.com/libsdl-org/SDL/files/15214259/SDL-sound-noise.zip)

```c
#include <stdio.h>
#include <SDL3/SDL.h>

int main(int, char **)
{
    if (SDL_Init(SDL_INIT_AUDIO) != 0) {
        printf("Error: SDL_Init(): %s\n", SDL_GetError());
        return 1;
    }

    const SDL_AudioSpec AudioSpec{ SDL_AUDIO_S16, 2, 48000 };

    SDL_AudioStream *pOutputStream = SDL_OpenAudioDeviceStream(SDL_AUDIO_DEVICE_DEFAULT_OUTPUT, &AudioSpec, nullptr, nullptr);
    if (pOutputStream == nullptr) {
        printf("Couldn't create audio playback stream. Error: %s", SDL_GetError());
        return 1;
    }
    SDL_ResumeAudioDevice(SDL_GetAudioStreamDevice(pOutputStream));

    SDL_AudioStream *pInputStream = SDL_OpenAudioDeviceStream(SDL_AUDIO_DEVICE_DEFAULT_CAPTURE, &AudioSpec, nullptr, nullptr);
    if (pInputStream == nullptr) {
        printf("Couldn't create audio capture stream. Error: %s", SDL_GetError());
        SDL_DestroyAudioStream(pOutputStream);
        return 1;
    }
    SDL_ResumeAudioDevice(SDL_GetAudioStreamDevice(pInputStream));

    uint8_t buffer[8192];
    while (true) {
        while (SDL_GetAudioStreamAvailable(pInputStream) > 0) {
            const int nRead = SDL_GetAudioStreamData(pInputStream, buffer, sizeof(buffer));
            if (nRead > 0) {
                if (0 != SDL_PutAudioStreamData(pOutputStream, buffer, nRead)) {
                    printf("Failed to put stream data, error %s", SDL_GetError());
                }
            }
        }

        if (false) {
            SDL_DestroyAudioStream(pInputStream);
            pInputStream = SDL_OpenAudioDeviceStream(SDL_AUDIO_DEVICE_DEFAULT_CAPTURE, &AudioSpec, nullptr, nullptr);
            SDL_ResumeAudioDevice(SDL_GetAudioStreamDevice(pInputStream));
        }

        SDL_Delay(10);
    }

    // not reachable
    SDL_DestroyAudioStream(pInputStream);
    SDL_DestroyAudioStream(pOutputStream);
    SDL_Quit();
    return 0;
}```

**Comment by icculus** [comment]
This fix got backed out, reopening this issue.

**Comment by 0x1F9F1** [comment]
The recording part of the fix was re-applied in 56e85064ffbc9afbcb7ab811f14d5439db26d7cf, though I do plan to tidy it up a bit, since I realised what was wrong with my previous fix (I think).

**Comment by slouken** [comment]
With the recording fix in https://github.com/libsdl-org/SDL/commit/56e85064ffbc9afbcb7ab811f14d5439db26d7cf, is this bug still active?

**Comment by joey-cyt** [comment]
> With the recording fix in [56e8506](https://github.com/libsdl-org/SDL/commit/56e85064ffbc9afbcb7ab811f14d5439db26d7cf), is this bug still active?

The original observed issue likely was gone with 56e8506. With the latest code, debug breaking does not be able to reproduce the noise issue, while previously it was reproducible with this method 100%. 
Every occasionally, I could hear a little bit noise but it backs to normal very soon. Not sure if this is related.

**Comment by slouken** [comment]
Okay, thanks!
